internal/character/clara/talent.go

```diff
@@ -0,0 +1,153 @@
+		Name:   TalentCounter,
+		Source: c.id,
+	})
+
+	c.engine.AddModifier(c.id, info.Modifier{
+		Name:   TalentRes,
+		Source: c.id,
+		Stats:  info.PropMap{prop.AllDamageReduce: 0.1},
+	})
+}
+
+// MAvatar_Klara_00_PassiveATK_Mark
+//
+// listens to AttackEnd events and check if clara is eligible for counter
+func (c *char) talentActionEndListener(e event.AttackEnd) {
+	attackerID := e.Attacker
+
+	// won't counter non-enemies (dominated ally)
+	if !c.engine.IsEnemy(attackerID) {
+		return
+	}
+
+	if c.canCounter(e) {
+		c.doCounter(attackerID)
+	}
+}
+
+// executes the follow-up attack,
+// NOTE: follow-up attack is same for both talent and e6
+func (c *char) doCounter(attackerID key.TargetID) {
+	c.engine.InsertAbility(info.Insert{
+		Execute: func() {
+			// DamageMaps
+			normalDamage := info.DamageMap{
+				model.DamageFormula_BY_ATK: talent[c.info.TalentLevelIndex()],
+			}
+			enhancedDamage := info.DamageMap{
+				model.DamageFormula_BY_ATK: talent[c.info.TalentLevelIndex()] +
+					ultDmgBoost[c.info.UltLevelIndex()],
+			}
+			mainTargetDamage := normalDamage
+
+			if c.engine.HasModifier(c.id, UltCounter) {
+				mainTargetDamage = enhancedDamage
+			}
+
+			// normal counter, damage
+			c.engine.Attack(info.Attack{
+				Source:       c.id,
+				Targets:      []key.TargetID{attackerID},
+				DamageType:   model.DamageType_PHYSICAL,
+				AttackType:   model.AttackType_INSERT,
+				BaseDamage:   mainTargetDamage,
+				StanceDamage: 30.0,
+				EnergyGain:   5.0,
+			})
+
+			// enhanced counter, damage to adjacent
+			// NOTE: in-game wording = damage value is based on main target's def
+			// mhy memes ?
+			if c.engine.HasModifier(c.id, UltCounter) {
+				splashDamage := info.DamageMap{model.DamageFormula_BY_ATK: (talent[c.info.TalentLevelIndex()] + ultDmgBoost[c.info.UltLevelIndex()]) * 0.5}
+
+				c.engine.Attack(info.Attack{
+					Source:       c.id,
+					Targets:      c.engine.AdjacentTo(attackerID),
+					DamageType:   model.DamageType_PHYSICAL,
+					AttackType:   model.AttackType_INSERT,
+					BaseDamage:   splashDamage,
+					StanceDamage: 30.0,
+				})
+			}
+		},
+		Source:     c.id,
+		Priority:   info.CharInsertAttackSelf,
+		AbortFlags: []model.BehaviorFlag{model.BehaviorFlag_STAT_CTRL, model.BehaviorFlag_DISABLE_ACTION},
+	})
+
+	// add marker modifier on enemy attacker
+	c.engine.AddModifier(attackerID, info.Modifier{
+		Name:   TalentMark,
+		Source: c.id,
+		State:  State{skillLevelIndex: c.info.SkillLevelIndex(), ultLevelIndex: c.info.UltLevelIndex()},
+	})
+}
+
+// helper fn to see if a counter is eligible
+//
+// - if clara is not targeted AND < E6 -> no counter
+//
+// - clara targeted OR E6 + winning E6 50/50 -> counter
+func (c *char) canCounter(e event.AttackEnd) bool {
+	isClara := false
+	for _, target := range e.Targets {
+		if target == c.id {
+			isClara = true
+		}
+	}
+	if isClara || (c.engine.HasModifier(c.id, E6) && c.engine.Rand().Float32() < 0.5) {
+		return true
+	}
+	return false
+}
```

Comment:
1. You need to check if the non-clara targets are an ally
2. the E6 50% chance check needs to happen **per ally hit** (so if 3 allies are hit, that is 50% fixed change 3 times)
3. Missing check for if ult is active/has count (**note:** ult does not decrement until insert is executed)
4. Check clara's eidolon level rather than using E6 modifier


Simplified version:
```go
func (c *char) canCounter(e event.AttackEnd) bool {
  ult := c.engine.HasModifier(c.id, UltCounter)
  for _, target := range e.Targets {
    if !c.engine.IsCharacter(target) {
      continue  
    }

    if target == c.id || ult || (c.info.Eidolon >= 6 && c.engine.Rand().Float64() < 0.5) {
      return true
    }
  }
  return false
}
```